### ISSUE TITLE
Lazy loading LuaSnip breaks lazy_load()ing of snippets in first buffer

### ISSUE THREAD

**Issue by marwing** [L3MON4D3/LuaSnip#434]
When lazy loading LuaSnip (in this case with packer, but shouldn't matter) and using `require('luasnip.loaders.from_snipmate').lazy_load()` (and probably the other lazy_load()s too) snippets don't get registered in the initial buffer if it doesn't have a FileType.

This probably comes from the [autocmds defined in luasnip.vim](https://github.com/L3MON4D3/LuaSnip/blob/4c06eecc1d47b239de58989ed826c862a2680da7/plugin/luasnip.vim#L46-L54) not triggering as BufWinEnter has already happened and the buffer doesn't have a FileType to trigger. Manually calling the appropriate function fixes the issue.

I would expect snippets to be available nonetheless without having to resort to non-lazy loading of either LuaSnip or snippets.
I unfortunately have no idea how to fix this myself.

**Comment by L3MON4D3** [comment]
Actually, the lua-loader might work because there we just load the lazy_load-snippets when `lazy_load` is called, which is an artifact from before we had the autocommands in `luasnip.vim`.
Well, we could provide a file that just registers the appropriate autocommands so luasnip can track which filetypes were already loaded.
Is it possible to load just a single file from a plugin even when lazy-loading it?
(an explicit `:luafile` with the path to that file will certainly work, would of course be nicer if there was a built-in way)

**Comment by marwing** [comment]
Thanks for the quick reply.
Loading non-lua snippets works by somehow (for example by `:set ft=`) triggering said autocmds after LuaSnip has been loaded.
Unfortunately I don't understand what you want me to to try. I have never used lua snippets and don't know how to use them.

**Comment by leiserfg** [comment]
How are you setting up luasnip in packer? are you lazyloading it?

**Comment by marwing** [comment]
I am using `event = 'InsertEnter'.
Lazy loading LuaSnip also does work mostly. Only when opening a file that doesn't have a filetype (no ftdetect script sets one) or a new buffer snippets don't appear until I set a filetype manually (can also be empty) or open a new buffer it loads snippets, as far as I understand (backed by testing) because both those actions trigger BufWinEnter or FileType events.

**Comment by L3MON4D3** [comment]
> Unfortunately I don't understand what you want me to to try

My idea would be to just add the autocommands on regular startup, then `lazy_load` the rest of luasnip on `InsertEnter`. Could you try that?
As snippets should be `lazy_load`ed, we cannot get around registering these autocommands on startup, but because they don't depend on anything else in luasnip explicitly, the impact on immediate startuptime should be negligible.

**Comment by marwing** [comment]
If that would be helpful I can create a minimal config? I am totally unfamiliar with luasnips internals and probably not too much of a help.

**Comment by marwing** [comment]
@L3MON4D3 Just adding the autocmds to my init doesn't help the situation. However if I understand the implementation of `lazy_load` and `_lazyload` from snipmate correctly the functions from the autocmds have to be called after the respective `lazy_load`s.
The autocmds get defined correctly so that is not the problem. The issue seems to be that when that when luasnip and my luasnip config gets loaded no event triggers the autocmds.

**Comment by leiserfg** [comment]
> I am, using `event = 'InsertEnter'. Lazy loading LuaSnip also does work mostly. Only when opening a file that doesn't have a filetype (no ftdetect script sets one) or a new buffer snippets don't appear until I set a filetype manually (can also be empty) or open a new buffer it loads snippets, as far as I understand (backed by testing) because both those actions trigger BufWinEnter or FileType events.

That's the problem, you are delaying the config so it's happening after the `BufWinEnter` so there is not much we can do (If we use another event someone else could do the same with that one in packer). Better not lazyload the plugin, you are not gaining any noticeable time with that.

**Comment by L3MON4D3** [comment]
> If that would be helpful I can create a minimal config?

Mhmm no, I think the problem is pretty clear :+1:

> the functions from the autocmds have to be called after the respective `lazy_load`s

Not necessarily, `lazy_load` loads already-loaded filetypes immediately.

> Just adding the autocmds to my init doesn't help the situation

Have you tried it in an empty buffer only? It should work as long as `BufWinEnter` is called upon loading the file.

For the empty Buffer, maybe we can change `BufWinEnter` to `BufNew`. I'll have to look into its semantics a bit more, but it sounds pretty good.

**Comment by L3MON4D3** [comment]
> > I am, using `event = 'InsertEnter'. Lazy loading LuaSnip also does work mostly. Only when opening a file that doesn't have a filetype (no ftdetect script sets one) or a new buffer snippets don't appear until I set a filetype manually (can also be empty) or open a new buffer it loads snippets, as far as I understand (backed by testing) because both those actions trigger BufWinEnter or FileType events.
> 
> That's the problem, you are delaying the config so it's happening after the `BufWinEnter` so there is not much we can do (If we use another event someone else could do the same with that one in packer). Better not lazyload the plugin, you are not gaining any noticeable time with that.

I agree, I think we can support it as long as only small changes are required, but anything beyond that just isn't worth it.

**Comment by L3MON4D3** [comment]
> Snippets are only missing when launching neovim into a buffer without a filetype. This can be an empty one or a file it doesn't recognize. Opening a file that does get recognized (has a filetype) works perfectly and snippets show up when entering insert mode.

Okay, nice, that confirms my understanding so far :+1:
I'm doing some changes on `lazy_load_support`: the autocommands are in a separate file now (`plugin/loader_autocommands.vim`), so you can just `:source` that on startup.

> BufNew doesn't change anything regarding this problem when I substitute BufWinEnter for it as a test.

Ah, shame. I've thought about it again, we cannot use `BufNew` because modelines aren't loaded yet, so we'd not load snippets for a ft specified in there.

> I do though. Somewhere between 5 and 20ms (in combination with lazy loading nvim-cmp) which is almost a fifth of my entire startup time.

Wouldn't say 5-20ms (less than that) is "noticeable", especially since that is just shifted to the first `InsertEnter`, but we can agree to disagree here, we should be able to support it in this instance :)

> This probably isn't a good longterm solution though.

True, I'm thinking of just loading `"all"` and the `""` (empty) filetype immediately when `lazy_load` is called, I'll look into that later.

**Comment by leiserfg** [comment]
Very rare, for me luasnip does not take more than `0.17 ms` to load.

**Comment by leiserfg** [comment]
> Luasnip itself doesn't take that long (`~0.13ms` for luasnip.vim) my luasnip config (according to `PackerProfile`) way longer than that. This matches most other lua plugins that don't automatically run a lot or even any code until the user calls into the plugins lua code (often a `setup()` function) which often takes way longer than just vim sourcing the plugin/ directory.

Well, that is true, calling `setup()` and `async_load()` takes longer but it's not much more than 5ms and is less error-prone.

**Comment by L3MON4D3** [comment]
I did some testing on my own, and somehow loading snippets for the blank,initial buffer works now, whereas I'm pretty sure it didn't a while ago (without changes on our end, I think)
For reference:
* check out luasnip/lazy_load_support
* call
  ```lua
  -- this path has to be adjusted, probably to ~/.local/share/nvim/site/pack/packer/start/luasnip/...
  vim.cmd("source /home/simon/Code/Lua/luasnip/plugin/loader_autocommands.vim")
  ```
  somewhere in your config

* ```lua
  use({
  	luasnip,
  	event = "InsertEnter",
        -- no other way for now, luasnip has to be loaded when the autocommands are executed, and luasnip is `require`d.
  	module = "luasnip.loaders.from_lua"
  })
  ```
and this always loads snippets correctly ("it works on my machine" :P)

**Comment by marwing** [comment]
Thanks for your effort on this (and LuaSnip in general), however this branch doesn't do anything useful for lazy loading.
It fixes the issue about missing snippets, however it does so by loading luasnip at startup because of the module lazy load by packer when the autocmds trigger making the `InsertEnter` event loading irrelevant.

I've decided not to lazy load luasnip for now and find my ms somewhere else as it doesn't seem worth the hassle. Especially since the only part not working is filetypeless buffers and only until a new buffer gets opened somehow (even a `:help` works).

If you don't have any other ideas feel free to close this.

**Comment by L3MON4D3** [comment]
> It fixes the issue about missing snippets, however it does so by loading luasnip at startup because of the module lazy load by packer when the autocmds trigger making the InsertEnter event loading irrelevant.

Well, not startup but `BufWinEnter`, but I can see that that isn't much better than startup.

The last way I can think of is allowing explicitly loading `lazy_load`-filetypes. This is implemented in `lazy_load_support`, just call `require("luasnip.loaders").load_lazy_loaded({"<ft1>", "<ft2>"})` somewhere in your config.

Calling it with just `{"", "all"}` will work for simple cases, but I wouldn't rely on it (packer does a `doautocmd BufRead` upon loading, which seems to also trigger a `BufWinEnter` (somehow? according to stacktrace), but only when the filetype isn't empty (hence only for `{"", "all"}`, other filetypes are loaded via that `BufWinEnter`)).

For a robust config, I'd recommend keeping track of loaded filetypes yourself (eg. via `BufWinEnter`) and calling `load_lazy_loaded` with those.

One more alternative, which is also simpler but more error-prone is calling `load_lazy_loaded` with the current filetype and `"all"`.

**Comment by L3MON4D3** [comment]
Closed with #435.

**Comment by marwing** [comment]
Thanks
